### PR TITLE
Unboxed vec128 mixed block fields

### PR DESCRIPTION
**Pull request by TheNumbat**
Fields of layout `vec128` are now allowed in mixed blocks. 

This is mostly a straightforward extension, except that during cmm conversion, `block_load` and `block_set` need to remap the field index to an offset in words. I've added `Mixed_block_shape.offset_in_words` for this purpose.

Additionally, flambda2 static mixed block generation has been updated to handle fields that take up more than one word.

Typing tests have had their output promoted; the unboxed simd test has been updated with vec128 fields.

### CODE REVIEW

**Review by ncik-roberts — COMMENTED** [review]
Frontend pieces look good. I'd approve but:
* someone else should take a second look at the backend / middle-end pieces.
* the tests are maybe a little light? Is there anything that should be added to `ocaml/testsuite/tests/mixed-blocks/test_mixed_blocks.ml` or `ocaml/testsuite/tests/mixed-blocks/constructor_args.ml`, or can we not use SIMD types in that testsuite?

**Comment by TheNumbat** [comment]
@ncik-roberts I've updated the native generated mixed blocks tests to use vec128s. However, it seems flambda2 static mixed block generation needs to be updated, will fix...

File: middle_end/flambda2/to_cmm/to_cmm_primitive.ml [hunk 1]
```diff
@@ -128,47 +129,60 @@ let make_block ~dbg kind alloc_mode args =
...
-      | Naked_float32 -> C.get_field_unboxed_float32
-      | Naked_int32 -> C.get_field_unboxed_int32
-      | Naked_int64 | Naked_nativeint -> C.get_field_unboxed_int64_or_nativeint)
+      C.get_field_computed Immediate, field
+    | Naked_floats _ -> C.unboxed_float_array_ref, field
+    | Mixed { field_kind = Flat_suffix field_kind; shape; _ } ->
+      let func =
+        match field_kind with
+        | Tagged_immediate -> C.get_field_computed Immediate
+        | Naked_float -> C.unboxed_float_array_ref
```
Comment (mshinwell) [inline middle_end/flambda2/to_cmm/to_cmm_primitive.ml:146]: Please leave a CR to make the naming of this function consistent, and likewise for "setfield_computed" below (should be "set_field_computed").

**Review by Gbury — COMMENTED** [review]
The middle end parts look good. Good to merge once my comment is answered.

File: middle_end/flambda2/to_cmm/to_cmm_static.ml [hunk 2]
```diff
@@ -248,25 +261,28 @@ let static_const0 env res ~updates (bound_static : Bound_static.Pattern.t)
...
       | Mixed_record shape ->
-        C.black_mixed_block_header tag num_fields
-          ~scannable_prefix_len:
-            (Flambda_kind.Mixed_block_shape.value_prefix_size shape)
+        ( MBS.field_kinds shape |> Array.to_list,
+          C.black_mixed_block_header tag (MBS.size_in_words shape)
+            ~scannable_prefix_len:(MBS.value_prefix_size shape) )
+    in
+    let static_fields =
+      List.concat_map (static_field res) (List.combine fields field_kinds)
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply middle_end/flambda2/to_cmm/to_cmm_static.ml]: Sure, fixed

File: middle_end/flambda2/to_cmm/to_cmm_static.ml [hunk 3]
```diff
@@ -385,7 +402,12 @@ let static_const0 env res ~updates (bound_static : Bound_static.Pattern.t)
   | Block_like s, Immutable_value_array fields ->
     let sym = R.symbol res s in
     let header = C.black_block_header 0 (List.length fields) in
-    let static_fields = List.concat_map (static_field res) fields in
+    let field_kinds =
+      List.init (List.length fields) (fun _ -> Flambda_kind.value)
+    in
+    let static_fields =
+      List.concat_map (static_field res) (List.combine fields field_kinds)
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply middle_end/flambda2/to_cmm/to_cmm_static.ml]: Fixed

File: middle_end/flambda2/to_cmm/to_cmm_static.ml [hunk 4]
```diff
@@ -24,7 +24,21 @@ module R = To_cmm_result
...
+
+  let concat_map2 f l1 l2 =
+    let rec aux f acc = function
+      | [], [] -> rev acc
+      | a1 :: l1, a2 :: l2 ->
+        let xs = f a1 a2 in
+        aux f (rev_append xs acc) (l1, l2)
+      | _, _ -> invalid_arg "List.concat_map2"
+    in
+    aux f [] (l1, l2)
```
Comment (Gbury) [inline middle_end/flambda2/to_cmm/to_cmm_static.ml:38]: This would better fit in `Misc.Stdlib.List` rather than here in this file.

**Review by ncik-roberts — APPROVED** [review]

**Pull request merged**